Puppeteer

d2fong opened the pull request:
migrate from phantomjs to puppeteer for the browser automation.

d2fong committed:
update bluebird

d2fong committed:
use puppeteer instead of phantomjs

maxkfranz commented:
TODO investigate why Travis is failing.  I think there may be two issues:

(1) The Travis VMs may not always have the libs necessary to run Chrome.  This is what you saw the other day on the Ubuntu VM, Dylan.

(2) I don't think Puppeteer works on old versions of Node.  We may have to drop support for versions before 6.

d2fong commented:
ubuntu steps for phantomjs:
https://gist.github.com/d2fong/8b7b48dd2c853034e5785bedf992e6aa

ubuntu steps for puppeteer:
https://github.com/GoogleChrome/puppeteer/blob/master/docs/troubleshooting.md#running-puppeteer-in-docker

jeshan commented:
I tried this in Docker on my local machine and it's working. Based on the previous comment, here is what I found that we need:

1. add packages as per troubleshooting doc above:
```dockerfile
RUN apt-get update && apt-get install -yq libgconf-2-4

# Install latest chrome dev package and fonts to support major charsets (Chinese, Japanese, Arabic, Hebrew, Thai and a few others)
# Note: this installs the necessary libs to make the bundled version of Chromium that Puppeteer
# installs, work.
RUN apt-get update && apt-get install -y wget --no-install-recommends \
    && wget -q -O - https://dl-ssl.google.com/linux/linux_signing_key.pub | apt-key add - \
    && sh -c 'echo "deb [arch=amd64] http://dl.google.com/linux/chrome/deb/ stable main" >> /etc/apt/sources.list.d/google.list' \
    && apt-get update \
    && apt-get install -y google-chrome-unstable fonts-ipafont-gothic fonts-wqy-zenhei fonts-thai-tlwg fonts-kacst ttf-freefont \
      --no-install-recommends \
    && rm -rf /var/lib/apt/lists/* \
    && apt-get purge --auto-remove -y curl \
    && rm -rf /src/*.deb


RUN groupadd -r pptruser && useradd -r -g pptruser -G audio,video pptruser \
    && mkdir -p /home/pptruser/Downloads \
    && chown -R pptruser:pptruser /home/pptruser \
    && chown -R pptruser:pptruser node_modules

# Run everything after as non-privileged user.
USER pptruser

```
2. add SYS_ADMIN capability. e.g in Docker compose:
```yaml
    cap_add:
      - SYS_ADMIN
```

I found that using puppeteer allowed for use of certain selectors, these were not possible in the current release with phantom (it crashed):
```js
{
  selector: 'edge[type = "Whatever"]',
  css: {
    'curve-style': 'unbundled-bezier',
....
```
I need the above selectors in my app, so hopefully we can get a release soon.

Note: tested on a node 11 based image. I will try to gather results based on older versions of Node

maxkfranz committed:
Clean up
- Add eslint
- Rename phantom dir to browser
- 'var' to 'let' in index.js
- Lint cleanup in index.js
- Remove lodash #20
- Update Cytoscape #21

maxkfranz committed:
Add support for external layout extensions

There is a mocha test that generates a PNG image for a DAG that uses the Dagre layout extension.  This is the same graph from the Dagre demo, and so the generated image is easy to verify.

maxkfranz committed:
Updates to readme re. extensions

maxkfranz committed:
Add test for cose layout

Ref #15

maxkfranz committed:
Remove generated index.js file

maxkfranz committed:
Update gitignore for cose test

maxkfranz committed:
Add `options.args` and other set up for Travis

maxkfranz committed:
Revise puppeteer links

maxkfranz commented:
@d2fong Let's test this on the interactions app next week.  I suspect we may need to expose the entire `options` object for `puppeteer.launch(options)`, since different setups may need to override more things.

d2fong commented:
Ok sounds good

d2fong commented on index.js:
```diff
@@ -106,51 +141,49 @@ proto.shot = function( opts, next ){
-    return page.property('viewportSize', { width: opts.width, height: opts.height });
+    return snap.browser.newPage();
+  }).then(function( puppeteerPage ){
+    page = puppeteerPage;
```
what do you think of caching page so that we don't have to create a newPage every time someone wants a snapshot?

https://github.com/GoogleChrome/puppeteer/issues/1569
https://github.com/GoogleChrome/puppeteer/issues/486

maxkfranz replied:
I wanted to maintain separation for two reasons:

(1) You can generate multiple images at once.

(2) It helps to avoid potential bugs.  For example, if there's a bug in Cytoscape when running for a long time with multiple stylesheets etc., then one page per output avoids it.  We're also setting a lot of globals in the page, which could be messy re. re-use.

The most important one is (1), whereas (2) is relatively unlikely.
